### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Basic

### PR DESCRIPTION
**Pull request by 4e554c4c**
---

- [x] depends on: #2105 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,2723 @@
...
+
+@[simp]
+theorem coeFn_generalLinearEquiv (f : GeneralLinearGroup R M) :
+    ⇑(generalLinearEquiv R M f) = (f : M → M) :=
+  rfl
+#align linear_map.general_linear_group.coe_fn_general_linear_equiv LinearMap.GeneralLinearGroup.coeFn_generalLinearEquiv
+
+end GeneralLinearGroup
+
+end LinearMap
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Basic.lean:2723]: Note that we can remove this if #18345 goes through first.

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Basic.lean:2745]: That PR is merged, so this _could_ be deleted

Comment (4e554c4c) [reply Mathlib/LinearAlgebra/Basic.lean]: done in [4bfe92c](https://github.com/leanprover-community/mathlib4/pull/1979/commits/4bfe92c44242d4272277abc381c99dcd20074989)

**Comment by jcommelin** [comment]
Blocked by lean4#2003 and the likes

**Comment by kbuzzard** [comment]
I am more optimistic (having looked at the first error): I think it's blocked by the fact that https://github.com/leanprover/lean4/pull/2093 hasn't made it to mathlib yet but in contrast to lean4#2003 this change has already made it to lean 4 master.

**Comment by jcommelin** [comment]
I can confirm that the first error goes away after merging the bump branch

**Comment by jcommelin** [comment]
This PR will compile once (some version of) #2105 is merged.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2105
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2105~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/LinearAlgebra/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,2680 @@
...
+theorem le_ker_iff_map [RingHomSurjective τ₁₂] {f : F} {p : Submodule R M} :
+    p ≤ ker f ↔ map f p = ⊥ := by rw [ker, eq_bot_iff, map_le_iff_le_comap]
+#align linear_map.le_ker_iff_map LinearMap.le_ker_iff_map
+
+theorem ker_codRestrict {τ₂₁ : R₂ →+* R} (p : Submodule R M) (f : M₂ →ₛₗ[τ₂₁] M) (hf) :
+    ker (codRestrict p f hf) = ker f := by rw [ker, comap_codRestrict, map_bot]; rfl
+#align linear_map.ker_cod_restrict LinearMap.ker_codRestrict
+
+theorem range_codRestrict {τ₂₁ : R₂ →+* R} [RingHomSurjective τ₂₁] (p : Submodule R M)
+    (f : M₂ →ₛₗ[τ₂₁] M) (hf) : range (codRestrict p f hf) = comap p.subtype (LinearMap.range f) := by
```
Comment (RemyDegenne) [inline Mathlib/LinearAlgebra/Basic.lean:1382]:
```suggestion
    (f : M₂ →ₛₗ[τ₂₁] M) (hf) :
    range (codRestrict p f hf) = comap p.subtype (LinearMap.range f) := by
```

**Comment by Komyyy** [comment]
As a temporary remedy, this command removes all red errors.
```
attribute [-instance] Ring.toNonAssocRing
```

File: Mathlib/LinearAlgebra/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,2698 @@
...
+
+-/
+
+
+open Function
+
+open BigOperators Pointwise
+
+-- Porting note: TODO Erase this line.
+attribute [-instance] Ring.toNonAssocRing
```
Comment (Komyyy) [inline Mathlib/LinearAlgebra/Basic.lean:68]: TODO: Erase this line.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**